tensorflow/compiler/xla/service/gpu/gpu_fused_mha_runner.cc

```diff
@@ -0,0 +1,520 @@
+      assign_seed();
+      break;
+    case CudnnfMHAKind::kScaleBiasSoftmax:
+      TF_RETURN_IF_ERROR(check_and_assign_bias());
+      break;
+    default:
+      return InternalError("Unknown fmha kind");
+  }
+  return config;
+}
+
+/*static*/ StatusOr<GpufMHAParams> GpufMHAParams::For(
+    const GpufMHAConfig &config, se::DeviceMemoryBase lhs_bmm1_buffer,
+    se::DeviceMemoryBase rhs_bmm1_buffer, se::DeviceMemoryBase rhs_bmm2_buffer,
+    se::DeviceMemoryBase output_buffer, se::DeviceMemoryBase mask_buffer,
+    se::DeviceMemoryBase bias_buffer) {
+  GpufMHAParams params;
+  params.config = &config;
+  params.lhs_bmm1_buffer = lhs_bmm1_buffer;
+  params.rhs_bmm1_buffer = rhs_bmm1_buffer;
+  params.rhs_bmm2_buffer = rhs_bmm2_buffer;
+  params.output_buffer = output_buffer;
+
+  auto assign_mask_buffer = [&]() {
+    params.mask_buffer.emplace();
+    se::DeviceMemoryBase &mask = *params.mask_buffer;
+    mask = mask_buffer;
+  };
+
+  auto assign_bias_buffer = [&]() {
+    params.bias_buffer.emplace();
+    se::DeviceMemoryBase &bias = *params.bias_buffer;
+    bias = bias_buffer;
+  };
+
+  switch (config.kind) {
+    case CudnnfMHAKind::kBmmBmm:
+    case CudnnfMHAKind::kSoftmaxDropout:
+    case CudnnfMHAKind::kSoftmax:
+      break;
+    case CudnnfMHAKind::kScaleMaskSoftmax:
+    case CudnnfMHAKind::kScaleMaskSoftmaxDropout:
+      TF_RET_CHECK(!mask_buffer.is_null());
+      assign_mask_buffer();
+      break;
+    case CudnnfMHAKind::kScaleBiasMaskSoftmax:
+    case CudnnfMHAKind::kScaleBiasMaskSoftmaxDropout:
+      TF_RET_CHECK(!mask_buffer.is_null());
+      TF_RET_CHECK(!bias_buffer.is_null());
+      assign_mask_buffer();
+      assign_bias_buffer();
+      break;
+    case CudnnfMHAKind::kScaleBiasSoftmax:
+    case CudnnfMHAKind::kScaleBiasSoftmaxDropout:
+      TF_RET_CHECK(!bias_buffer.is_null());
+      assign_bias_buffer();
+  }
+  return params;
+}
+
+Status RunGpuFMHA(
+    const GpufMHAConfig &fmha_config, se::DeviceMemoryBase lhs_bmm1_buffer,
+    se::DeviceMemoryBase rhs_bmm1_buffer, se::DeviceMemoryBase rhs_bmm2_buffer,
+    se::DeviceMemoryBase output_buffer, se::DeviceMemoryBase scratch_buffer,
+    se::DeviceMemoryBase mask_buffer, se::DeviceMemoryBase bias_buffer,
+    se::Stream *stream, RunFusedMHAOptions options) {
+  TF_ASSIGN_OR_RETURN(
+      GpufMHAParams params,
+      GpufMHAParams::For(fmha_config, lhs_bmm1_buffer, rhs_bmm1_buffer,
+                         rhs_bmm2_buffer, output_buffer, mask_buffer,
+                         bias_buffer));
+  PrimitiveType input_primitive_type = fmha_config.input_type;
+  switch (input_primitive_type) {
+    case F16:
+      return RunGpuFMHAImpl<Eigen::half, Eigen::half, Eigen::half>(
+          params, stream, scratch_buffer, options);
+    case BF16:
+      return RunGpuFMHAImpl<Eigen::bfloat16, Eigen::bfloat16, Eigen::bfloat16>(
+          params, stream, scratch_buffer, options);
+    default:
+      return Unimplemented("Unimplemented fused MHA");
```

Comment:
Provide more details which case is not supported and why? Maybe print the config?

Reply:
Addressed in 3453909b0099b3c09864ecb71763e5caa6ffd4ac